### PR TITLE
feat: update generate password functionality

### PR DESCRIPTION
**Pull request by vamsimundra**
What does this PR do?
Changed the functionality of password:generate command. Now generating the password with the Profile Password policies and Organization Password Policies

What issues does this PR fix or reference?
@W-5829533@

### CODE REVIEW

File: src/user.ts [hunk 1]
```diff
@@ -211,20 +231,48 @@ export class User extends AsyncCreatable<User.Options> {
...
+  public static generatePasswordUtf8(
+    passwordCondition: PasswordConditions = { length: 10, complexity: 1 }
+  ): SecureBuffer<void> {
     const ALLCHARS = UPPER.concat(LOWER, NUMBERS, SYMBOLS.join(''));
-    const password = [UPPER[rand(UPPER)], LOWER[rand(LOWER)], NUMBERS[rand(NUMBERS)], SYMBOLS[rand(SYMBOLS)]]
-      .concat(
-        Array(PASSWORD_LENGTH - 4)
+    let password: string[] = [];
+
+    // Required number of Lower letter characters. one character of each of the 4 types followed by the remaining random characters, then quasi-randomize
```
Comment (mshanemc) [inline src/user.ts:244]: overall, this feels like it could be simpler with booleans.  They all require 1 or 0, nothing higher than 1.

So, for example
```js
if (PASSWORD_COMPLEXITY[passwordCondition.complexity].LOWER) {
    password.push(LOWER[rand(LOWER)])
}
```

Comment (mshanemc) [reply src/user.ts:244]: another idea would be to iterate the types to avoid the repetition, something like

```ts
[SYMBOLS,NUMBERS,UPPER,LOWER].forEach( charSet => {
    if (PASSWORD_COMPLEXITY[passwordCondition.complexity][charSet]) {
        password.push(charType[rand(charSet)])
    }
})
```

Comment (mshanemc) [reply src/user.ts:244]: finally, we could generate the right length of password from the start (`length` randow lowercase characters) then overwrite it with whatever other characters are requied and then shuffle that.

Then there's a chance (for cases 0 and 1) where no changes have to be made at all (all the requirements were false)

File: src/user.ts [hunk 2]
```diff
@@ -29,14 +29,29 @@ import { SecureBuffer } from './secureBuffer';
...
 
+interface Complexity {
+  LOWER: number;
+  UPPER: number;
+  NUMBERS: number;
+  SYMBOLS: number;
+}
+
+const PASSWORD_COMPLEXITY: { [index: string]: Complexity } = {
+  '0': { LOWER: 0, UPPER: 0, NUMBERS: 0, SYMBOLS: 0 },
```
Comment (mshanemc) [inline src/user.ts:47]: definitely feels like booleans

File: src/user.ts [hunk 3]
```diff
@@ -29,14 +29,29 @@ import { SecureBuffer } from './secureBuffer';
...
-const PASSWORD_LENGTH = 13;
 const LOWER = 'abcdefghijklmnopqrstuvwxyz';
 const UPPER = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
 const NUMBERS = '1234567890';
 const SYMBOLS = ['!', '@', '#', '$', '%', '^', '&', '*', '(', ')', '_', '[', ']', '|', '-'];
 
 const rand = (len: Many<string>): number => Math.floor(Math.random() * len.length);
 
+interface Complexity {
+  LOWER: number;
```
Comment (mshanemc) [inline src/user.ts:40]: you could also make these all optional unless true.  So `1` could be { LOWER: true } and 5 would be { LOWER:true, NUMBERS:true, SYMBOLS:true }

File: messages/user.json [hunk 4]
```diff
@@ -1,9 +1,10 @@
...
-    "problemsDescribingTheUserObject": "Problems occurred while attempting to describe the user object",
-    "permsetNamesAreRequired": "The permission set names are missing but required."
+  "orgRequired": "An org instance is required.",
+  "userQueryFailed": "Failed to query for the user %s.",
+  "invalidHttpResponseCreatingUser": "An invalid http response code (%s) was received while trying to create a user.",
+  "missingFields": "The fields parameters is undefined.",
+  "missingId": "The Salesforce id for the user is required.",
+  "problemsDescribingTheUserObject": "Problems occurred while attempting to describe the user object",
+  "permsetNamesAreRequired": "The permission set names are missing but required.",
+  "complexityOutOffBound": "INVALID_COMPLEXITY: valid values are between 0 and 5"
```
Comment (WillieRuemmele) [inline messages/user.json:9]: Do we need to include `INVALID_COMPLEXITY` can we make it less server-like, and more friendly? Have you worked with Juliet on this message?

File: src/user.ts [hunk 5]
```diff
@@ -211,20 +234,30 @@ export class User extends AsyncCreatable<User.Options> {
...
+      throw SfdxError.create('@salesforce/core', 'user', 'complexityOutOffBound');
+    }
+    let password: string[] = [];
+    ['SYMBOLS', 'NUMBERS', 'UPPER', 'LOWER'].forEach((charSet) => {
+      if (PASSWORD_COMPLEXITY[passwordCondition.complexity][charSet]) {
+        password.push(CHARACTERS[charSet][rand(CHARACTERS[charSet])]);
+      }
+    });
 
+    // Concatinating remaining length randomly with all available characters
```
Comment (mshanemc) [inline src/user.ts:254]: the comment is no longer true (it's just going to be lowers).

File: messages/user.json (hunk 4 above)
Comment (mshanemc) [inline messages/user.json:9]:
```suggestion
  "complexityOutOfBound": "INVALID_COMPLEXITY: valid values are between 0 and 5"
```

Comment (mshanemc) [reply messages/user.json:9]: and also get Juliet to check the wording/format of this message.  The rest of the messages are complete sentences ending in a `.` so try to make it like that.

File: test/unit/userTest.ts [hunk 6]
```diff
@@ -127,6 +127,32 @@ describe('User Tests', () => {
...
+    it('Should generate length 12 and complexity 1 password containing at least one number and one  lowercase letter ', () => {
+      const passwordCondition = { length: 12, complexity: 1 };
+      const password: SecureBuffer<void> = User.generatePasswordUtf8(passwordCondition);
+      password.value((buffer: Buffer): void => {
+        const passwordAsArrayOfCharacters = buffer.toString('utf8').split('');
+        expect(passwordAsArrayOfCharacters.length).to.be.equal(12);
+        expect(
+          passwordAsArrayOfCharacters.some((char) => ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9'].includes(char))
+        );
+        expect(passwordAsArrayOfCharacters.some((char) => char.toLowerCase() !== char.toLowerCase()));
```
Comment (WillieRuemmele) [inline test/unit/userTest.ts:140]: is `char.toLowerCase() !== char.toLowerCase()` correct? seems like you're comparing the same thing to itself

File: test/unit/userTest.ts [hunk 7]
```diff
@@ -127,6 +127,32 @@ describe('User Tests', () => {
...
     });
+
+    it('Should generate length 12 and complexity 1 password containing at least one number and one  lowercase letter ', () => {
+      const passwordCondition = { length: 12, complexity: 1 };
+      const password: SecureBuffer<void> = User.generatePasswordUtf8(passwordCondition);
+      password.value((buffer: Buffer): void => {
+        const passwordAsArrayOfCharacters = buffer.toString('utf8').split('');
+        expect(passwordAsArrayOfCharacters.length).to.be.equal(12);
+        expect(
+          passwordAsArrayOfCharacters.some((char) => ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9'].includes(char))
```
Comment (mshanemc) [inline test/unit/userTest.ts:138]: regex `\d` ?

File: test/unit/userTest.ts [hunk 8]
```diff
@@ -127,6 +127,32 @@ describe('User Tests', () => {
...
+      });
+    });
+    it('Should generate length 14 and complexity 3 password containing at least one number and one  lowercase letter ', () => {
+      const passwordCondition = { length: 12, complexity: 1 };
+      const password: SecureBuffer<void> = User.generatePasswordUtf8(passwordCondition);
+      password.value((buffer: Buffer): void => {
+        const passwordAsArrayOfCharacters = buffer.toString('utf8').split('');
+        expect(passwordAsArrayOfCharacters.length).to.be.equal(12);
+        expect(
+          passwordAsArrayOfCharacters.some((char) => ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9'].includes(char))
```
Comment (mshanemc) [inline test/unit/userTest.ts:150]: this could be cleaner with regex, too.

File: test/unit/userTest.ts [hunk 9]
```diff
@@ -127,6 +127,32 @@ describe('User Tests', () => {
...
+        );
+        expect(passwordAsArrayOfCharacters.some((char) => char.toLowerCase() !== char.toLowerCase()));
+        expect(passwordAsArrayOfCharacters.some((char) => char.toLowerCase() !== char.toLowerCase()));
```
Comment (WillieRuemmele) [inline test/unit/userTest.ts:153]: The test says it is testing length 14, complexity 3. But the test is testing 12/1

File: test/unit/userTest.ts [hunk 10]
```diff
@@ -127,6 +127,32 @@ describe('User Tests', () => {
...
+    it('Should generate length 14 and complexity 3 password containing at least one number and one  lowercase letter ', () => {
+      const passwordCondition = { length: 12, complexity: 1 };
+      const password: SecureBuffer<void> = User.generatePasswordUtf8(passwordCondition);
+      password.value((buffer: Buffer): void => {
+        const passwordAsArrayOfCharacters = buffer.toString('utf8').split('');
+        expect(passwordAsArrayOfCharacters.length).to.be.equal(12);
+        expect(
+          passwordAsArrayOfCharacters.some((char) => ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9'].includes(char))
+        );
+        expect(passwordAsArrayOfCharacters.some((char) => char.toLowerCase() !== char.toLowerCase()));
```
Comment (WillieRuemmele) [inline test/unit/userTest.ts:152]: same as above

File: src/user.ts [hunk 11]
```diff
@@ -29,14 +29,32 @@ import { SecureBuffer } from './secureBuffer';
...
+  SYMBOLS: ['!', '@', '#', '$', '%', '^', '&', '*', '(', ')', '_', '[', ']', '|', '-'],
+};
+
+const PASSWORD_COMPLEXITY: { [index: number]: Complexity } = {
+  '0': { LOWER: true },
+  '1': { LOWER: true, NUMBERS: true },
+  '2': { LOWER: true, SYMBOLS: true },
+  '3': { LOWER: true, UPPER: true, NUMBERS: true },
+  '4': { LOWER: true, NUMBERS: true, SYMBOLS: true },
+  '5': { LOWER: true, UPPER: true, NUMBERS: true, SYMBOLS: true },
```
Comment (WillieRuemmele) [inline src/user.ts:55]: is 4 or 5 the most complex? We need to standardize on the server options

File: messages/user.json (hunk 4 above)
Comment (jshackell-sfdc) [inline messages/user.json:9]:
```suggestion
  "complexityOutOffBound": "Invalid complexity value. Valid values are between 0 and 5. "
```

**Review by mshanemc — CHANGES_REQUESTED** [review]
the function seems good, message needs help.

Thinking about the tests...definitely add one to test the complexity boundary error (ex: `9`).

what are the boundaries of the password length?  n >=8 ?  Writing a test with length=6 that you expect to fail would be a good.

**Review by WillieRuemmele — COMMENTED** [review]
Functionality looks good, tests could use some work

File: messages/user.json (hunk 4 above)
Comment (jshackell-sfdc) [reply messages/user.json:9]: Also is there a reason this error message is in sfdx-core and not plugin-user, together with lengthOutOfBound?  If possible I think it's better to keep them all together, unless there's a good reason not to.

Finally, maybe name the variable complexityOutOfBound? (just one "f").

File: src/user.ts (hunk 11 above)
Comment (jshackell-sfdc) [reply src/user.ts:55]: Hmm, Vamsi originally had 4 and 5 switched (at least that's what he showed me when I worked on the description) and I asked the same question!  I think he switched it.  But did I steer him wrong?

Comment (WillieRuemmele) [reply src/user.ts:55]: I think it makes more logical sense to have higher number = more security, but I thought the server had 4 as the most secure option... we need to make a note of what we're doing, and what the server is expecting

File: test/unit/userTest.ts [hunk 12]
```diff
@@ -127,6 +127,36 @@ describe('User Tests', () => {
...
+        const complexity3Regex = new RegExp('^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.{14})');
+
+        expect(complexity3Regex.test(passwordAsArrayOfCharacters));
+      });
+    });
+
+    it('Should throw an error beacuse of complexity', () => {
+      try {
+        const passwordCondition = { length: 14, complexity: 9 };
+        User.generatePasswordUtf8(passwordCondition);
```
Comment (WillieRuemmele) [inline test/unit/userTest.ts:155]: you need to add something like `assert.fail('the above should throw an error')` below `generatePasswordUtf8()` so incase the function doesn't throw an error, the test will fail

File: src/user.ts (hunk 11 above)
Comment (jshackell-sfdc) [reply src/user.ts:55]: @vamsimundra , please go with what the devs are saying!  Like I said earlier, to me the order you originally showed me looked wrong, which is why I flagged it, but I should really stay out of it.  But please make sure that whatever you decide is accurately reflected in the --help output.  Thanks

File: src/user.ts [hunk 13]
```diff
@@ -211,20 +234,32 @@ export class User extends AsyncCreatable<User.Options> {
...
-          .map(() => ALLCHARS[rand(ALLCHARS)])
-      )
-      .sort(() => Math.random() - 0.5);
+  public static generatePasswordUtf8(
+    passwordCondition: PasswordConditions = { length: 13, complexity: 5 }
+  ): SecureBuffer<void> {
+    if (!PASSWORD_COMPLEXITY[passwordCondition.complexity]) {
+      throw SfdxError.create('@salesforce/core', 'user', 'complexityOutOfBound');
+    }
+    if (passwordCondition.length < 8 || passwordCondition.length > 1000)
```
Comment (WillieRuemmele) [inline src/user.ts:247]: NIT: feel free to ignore this, but can we add {} for the second if statement? can we also add a new line between the error handling/throwing section, and the `let password`?

**Review by WillieRuemmele — APPROVED** [review]

**Review by mshanemc — APPROVED** [review]
